packages/astro/src/build/bundle/js.ts

```diff
@@ -21,6 +24,159 @@ export function collectJSImports(buildState: BuildOutput): Set<string> {
   return imports;
 }
 
+function pageUrlToVirtualJSEntry(pageUrl: string) {
+  return 'astro-virtual:' + pageUrl.replace(/.html$/, '').replace(/^\./, '') + '.js';
+}
+
+export async function bundleHoistedJS({
+  buildState,
+  astroConfig,
+  logging,
+  depTree,
+  dist,
+  runtime
+}: {
+  astroConfig: AstroConfig;
+  buildState: BuildOutput;
+  logging: LogOptions;
+  depTree: BundleMap;
+  dist: URL;
+  runtime: AstroRuntime;
+}) {
+  const sortedPages = Object.keys(depTree); // these were scanned in parallel; sort to create somewhat deterministic order
+  sortedPages.sort((a, b) => a.localeCompare(b, 'en', { numeric: true }));
+  
+  /**
+   * 1. Go over sorted pages and create a virtual module for all of its dependencies
+   */
+  const entryImports: string[] = [];
+  const virtualScripts = new Map<string, ScriptInfo>();
+  const pageToEntryMap = new Map<string, string>();
+
+  for(let pageUrl of sortedPages) {
+    const hoistedJS = depTree[pageUrl].hoistedJS;
+    if(hoistedJS.size) {
+      for(let [url, scriptInfo] of hoistedJS) {
+        if(virtualScripts.has(url) || !url.startsWith('astro-virtual:')) continue;
+        virtualScripts.set(url, scriptInfo);
+      }
+      const entryURL = pageUrlToVirtualJSEntry(pageUrl);
+      const entryJS = Array.from(hoistedJS.keys()).map(url => `import '${url}';`).join('\n');
+      virtualScripts.set(entryURL, {
+        content: entryJS
+      });
+      entryImports.push(entryURL);
+      pageToEntryMap.set(pageUrl, entryURL);
+    }
+  }
+
+  if(!entryImports.length) {
+    // There are no hoisted scripts, bail
+    return;
+  }
+
+  /**
+  * 2. Run the bundle to bundle each pages JS into a single bundle (with shared content)
```

Comment:
There was an original idea of Astro that we were moving from "bundle JS by page" to "bundle JS by each hydrated component", using the hydrated components as "entrypoints"/"inputs" for your final bundled output.

This seems to be breaking that idea a bit, and moving into a page-based bundling world. Can we treat hoisted scripts the same way that we treat hydrated JS components, and just handle them in our current bundle step? It feels odd to be doing two different JS bundling passes (three if you count CSS bundling) when every other tool that I'm familiar with would just do a single pass.

Reply:
We could, but that's not how CSS works and was not the original ask in #370. What would be the advantage to having N script tags in this case?

One disadvantage to doing it the other way I see (aside from more requests) is that the current implementation results in hashed bundle names and the component bundling does not.

Reply:
One advantage to doing the other way is consistency, just to be clear that I'm not completely opposed to it. I could see the argument for leaving it the other way (even though I don't think that's what users want and not how CSS works) and then proposing a change via RFC.

Reply:
I don't think I understand enough about what this code is doing to answer well, will leave for you and @natemoo-re to decide the right path forward. Happy to jump in a convo with you both to discuss more, as well!

Reply:
The reason for the 2 passes here is because the 2nd pass happens after the HTML has already been written and this bundling needs to update the HTML (to remove script tags and update to the hashed file names). In the future if we refactor the build, it should be to do all of the JS during this pass. Even though this is 2 passes, it's not recompiling the same code, these are different entry points.